fix an environment problem in merge_constraint

gasche opened the pull request:
This is a preliminary fix for an uncaught `Not_found` exception in the type-checker, from my analysis of the reproduction case provided by @hhugo in https://github.com/ocaml/ocaml/pull/9609#issuecomment-636710146 .

The patch includes a reproduction case:

```ocaml
module type Sig = sig
  type 'a ind = 'a * int
  type t = T : 'e ind -> t [@@unboxed]
end

module type Problem = sig
  include Sig
  module type ReboundSig = Sig
    with type 'a ind = 'a ind
     and type t = t
end
```

Without the patch, checking `type t = t` (which results in a recomputation of the validity of the `[@@unboxed]` attribute in the resulting signature) raises a `Not_found` exception during the lookup of a `Tconstr` path in the typing environment. (The code assumes that all type-constructor paths are bound in the typing environment.)

The patch fixes the issue by bluntly replacing a use of `initial_env` in `merge_constraint` by `env`. I suspect however that the patch is incorrect or incomplete. The authors of `merge_constraint` took care to separate the initial and partially-checked environments, I suppose there is a reason for it -- I would welcome an explanation. If using `env` was the correct change, then many other occurrences of `initial_env` should be modified.

(cc @garrigue and, somewhat randomly, @lpw25  @trefis @Octachron)

trefis commented:
I hope more knowledgeable people than me will offer an explanation, but after having looked briefly at the code it seems that we're indeed always "translating" the rhs of the constraint in the initial environment (i.e. the one just before the definition of the module type), and then checking that it is well-formed in the current environment.

That sort of makes sense to me, it's forcing the rhs of the constraint to refer only to things outside of the module type.

What I believe your patch does (if it was generalized to all the calls of `transl_with_constraint`) is to allow references to items that come before the one being constrained. I'm not sure that something we'd want.

Now, what's really surprising me is that in your example, `t` is supposed to exist in `initial_env`. So where does the `Not_found` come from? (Furthermore, my impression is that the new `t` shouldn't exist in `env` yet. But I probably misread that part)

gasche commented:
Being a bit imaginative, here is a proposed explanation (but I am not sure it matches the code): `mod with type t = foo and type u = bar` is evaluated by iterating two calls to `merge_constr` (in the `Pmty_with` of `transltype_mod_aux`), as `(mod with type t = foo) with type u = bar`. During the first call (`with t = foo`), we end up with a signature where `t` is replaced by a *new* type declaration that is not part of the environment. So then when we process the second constraint `with type u = bar`, if `bar` mentions `t` we have a problem.

lpw25 commented:
I agree with @trefis. `t` and `ind` should both be in the initial environment. The type definitions that the `with` constraints are applying to are in `Sig` and should be unrelated to the ones that are being looked up that come from the preceding `include`.

gasche commented:
Here is a reproduction case for a similar problem with variance instead of separability:

```ocaml
module type Sig = sig
  type +'a abstract
  type +'a user = Foo of 'a abstract
end

module type Problem = sig
  include Sig
  module M : Sig
    with type 'a abstract = 'a abstract
     and type 'a user = 'a user
  type +'a foo = 'a M.user
end
```

```
11 |   type +'a foo = 'a M.user
       ^^^^^^^^^^^^^^^^^^^^^^^^
Error: In this definition, expected parameter variances are not satisfied.
       The 1st type parameter was expected to be covariant,
       but it is injective invariant.
```

gasche commented:
I was amused to trace the introduction of the `initial_env` vs. `env` usage back to @xavierleroy's https://github.com/ocaml/ocaml/commit/df42185ca00941330082408f5eb7d2c2519fcb97 in 1996. Yet another person who may be able to shine light on this issue.

xavierleroy commented:
>  @xavierleroy's df42185 in 1996. Yet another person who may be able to shine light on this issue.

24 years later, I'm afraid not.  But @trefis 's explanation above (https://github.com/ocaml/ocaml/pull/9623#issuecomment-636786665) feels very plausible to me: the right-hand sides must be defined in the initial environment but must still make sense when brought into the new signature.  How this explains the loss of variance of separability info,  I don't know.

gasche committed:
add a testcase for 'with type <lhs> = <rhs>' scoping

gasche committed:
Typemod.merge_constraint: be careful about the typing environments

This PR fixes an old bug in the interaction between [merge_constraint]
and [Typedecl.transl_with_constraint], where
variance (and now separability) are recomputed in an invalid type
environment. See #9624 and the new tests.

gasche commented:
I now propose a fix that feels more correct and more complete to me. Reviews would be very welcome -- this bug is causing a regression in at least the Jane Street codebase for 4.11.

There was something fundamentally flawed with the idea that we would type-check a definition in an environment, but then we could only check its well-formedness in a richer environment.

I think that the problem was that the `initial_env` should be used to type-check the right-hand-side of the constraint, but not to recompute declaration properties which depend on `type_kind` (in particular the variant/record declaration) which is only valid in the signature environment `env`. The proposed patch passes both environments to `Typedecl.transl_with_constraint`, and tries to use either one when appropriate.

garrigue commented:
> I think that the problem was that the `initial_env` should be used to type-check the right-hand-side of the constraint, but not to recompute declaration properties which depend on `type_kind` (in particular the variant/record declaration) which is only valid in the signature environment `env`. The proposed patch passes both environments to `Typedecl.transl_with_constraint`, and tries to use either one when appropriate.

I'm not sure I understand this reasoning.
The right-hand side of a constraint is defined only using the original environment.
So I don't see why you would need to refer to the signature environment in it.
Basically, in a constraint you cannot refer to anything from the signature you are modifying.

Also, your scenario doesn't seem to make sense: even if the signature contains a type `t`, it will not be seen from the right-hand sides. The only place where the two environments interact is `check_type_decl`, but it only checks subsumption, and is already called with the right environment (`env` subsuming `initial_env` for paths).

I agree that your variance code demonstrates a bug, and I'm going to look into this.

lpw25 commented:
@garrigue The problem that @gasche is getting at is that we pass not just the right-hand side of the constraint but also the definition from the signature to `transl_with_constraint` as `orig_decl`. The `type_kind` and `type_param` fields from that declaration are both used with operations that depend on the environment -- so they either need to be given the inner environment of the signature, or we need to stop using them.

lpw25 commented:
I should probably also point out that the "internal" environment calculated by `merge_constraint` is also incorrect as it ignores recursion. The big red flag is that it calls `Env.add_item` which is always a mistake. You'll probably find that:
```ocaml
module type Sig = sig
  type +'a user = Foo of 'a abstract
  and +'a abstract
end

module type Problem = sig
  include Sig
  module M : Sig
    with type 'a abstract = 'a abstract
     and type 'a user = 'a user
  type +'a foo = 'a M.user
end
```
is broken even with your patch.

garrigue commented on typing/typedecl.ml:
```diff
@@ -1370,53 +1370,68 @@ let transl_value_decl env loc valdecl =
+  let arity_ok = List.length params = sig_decl.type_arity in
   if arity_ok then
-    List.iter2 (Ctype.unify_var env) params orig_decl.type_params;
+    List.iter2 (Ctype.unify_var outer_env) params sig_decl.type_params;
```
Actually, I think this should be `sig_env` rather than `outer_env`.
However, this means that this unification should only happen after all the `transl_simpl_type` (which should not refer to types of the signature), and protect against unification errors.

garrigue commented on typing/typedecl.ml:
```diff
@@ -1426,63 +1441,64 @@ let transl_with_constraint env id row_path orig_decl sdecl =
-          env ~check:true decl (Typedecl_variance.variance_of_sdecl sdecl)
+    let required = Typedecl_variance.variance_of_sdecl sdecl in
+    try
+      Typedecl_variance.compute_decl sig_env ~check:true new_sig_decl required
```
You are right: since the `type_kind` part comes from the signature, we need to use `sig_env` here.

garrigue commented on typing/typedecl.ml:
```diff
@@ -1426,63 +1441,64 @@ let transl_with_constraint env id row_path orig_decl sdecl =
+    Typedecl_immediacy.compute_decl sig_env new_sig_decl in
   let new_type_separability =
-    try Typedecl_separability.compute_decl env decl
+    try Typedecl_separability.compute_decl sig_env new_sig_decl
```
And also for immediacy and separability, of course.

garrigue reviewed (commented):
It took me a long while to understand the problem, but now I see it.
I'm going to try to build an example showing the problem with constrained parameters

garrigue commented:
> I should probably also point out that the "internal" environment calculated by `merge_constraint` is also incorrect as it ignores recursion. The big red flag is that it calls `Env.add_item` which is always a mistake.

I see the problem. In this case this could be solved by using an environment with the full signature, rather than expanding it incrementally. In general, like for functors, the variance (and other mutually recursive properties) should be computed on the final signature, after all substitutions are done. But maybe this last part can be delayed until we do it for functors.

gasche commented:
@lpw25 I understand what you mean and I might see how to go in a direction of fixing this problem (rewrite `merge_constraint` to work on each recursive group separately), but it seems orthogonal to the aspect that I am attacking in the present PR (I propose to pass the signature environment to the call, and you propose to compute a better signature environment). Would you like to propose a separate fix for that part?

garrigue commented:
Example of problem with `type_params`:
```ocaml
module type Sig2 =
  sig type 'a u = 'a list type +'a t constraint 'a = 'b u end
type +'a t = 'b constraint 'a = 'b list
module type Sig3 = Sig2 with type 'a t = 'a t
```

garrigue commented:
@gasche The problem @lpw25 describes is that, if you're calling `transl_with_constraint` with an incomplete environment, you may still get the same kind of errors for mutual recursion. So I would say this is part of the same bug. But this can be easily fixed by first computing an environment extended with whole signature, and then incrementally overwriting it with the constrained definitions. I.e., you just need to add one line in `merge_constraints`.

garrigue commented on typing/typedecl.ml:
```diff
@@ -1401,22 +1413,30 @@ let transl_with_constraint env id row_path orig_decl sdecl =
+     there. *)
+  let env = sig_env in
+  if arity_ok then
+    List.iter2 (Ctype.unify_var env) params sig_decl.type_params;
```
Here you may get a unification error.

gasche replied:
I don't understand what you mean. Was it not possible to get a unification error before? Should we do something specific about unification errors, and if so, why?

garrigue replied:
This is a question of unification order: since the `ptype_params` have to be distinct type variables, you cannot get a unification error by unifying them with constrained parameters. However, this could cause an error when unifying the `ptype_cstrs`, which would be reported as an `Inconsistent_constraint` error. So you need to wrap this call to `unify` in the same way.

gasche commented:
@garrigue I pushed a new commit that changes the place where the `type_params` are unified, and in fact reduces the diff a bit thanks to (judicious? dangerous?) use of a shadowed `env` variable. Let me know what you think.

I needed to tweak your example slightly to get it to compile in my fixed version. As it is written, with my commit it fails with the following error:

```
module ParamsUnificationEnv = struct
   module type Sig =
    sig type 'a u = 'a list type +'a t constraint 'a = 'b u end
   type +'a t = 'b constraint 'a = 'b list
   module type Sig2 = Sig with type 'a t = 'a t
 end
 [%%expect{|
Line 5, characters 21-46:
5 |   module type Sig2 = Sig with type 'a t = 'a t
                         ^^^^^^^^^^^^^^^^^^^^^^^^^
Error: In this `with' constraint, the new definition of t
       does not match its original definition in the constrained signature:
       Type declarations do not match:
         type 'a t = 'a t constraint 'a = 'b u
       is not included in
         type +'a t constraint 'a = 'b u
       Their variances do not agree.
|}]
```

Two changes make this compile: I can drop the variance annotations

```
  module type Sig =
    sig type 'a u = 'a list type 'a t constraint 'a = 'b u end
  type 'a t = 'b constraint 'a = 'b list
  module type Sig2 = Sig with type 'a t = 'a t
```

or I can define `type +'a t = b` in the first Signature as well:

```
  module type Sig =
    sig type 'a u = 'a list type +'a t = 'b constraint 'a = 'b u end
  type +'a t = 'b constraint 'a = 'b list
  module type Sig2 = Sig with type 'a t = 'a t
```

garrigue reviewed (commented):
Additionally, in `merge_constraint`, I would add
```ocaml
    let full_env = Env.add_signature sg initial_env in
    let (tcstr, sg) = merge full_env sg names None in
```

gasche commented:
Hm, it also works with `with type +'a t = 'a t`.

```
  module type Sig =
    sig type 'a u = 'a list type +'a t constraint 'a = 'b u end
  type +'a t = 'b constraint 'a = 'b list
  module type Sig2 = Sig with type +'a t = 'a t
```

lpw25 commented:
@gasche My proposed solution would be what Jacques said: use `Env.add_signature` before entering a signature and remove the call to `Env.add_item`.

> Hm, it also works with with type +'a t = 'a t

Does that actually check the variance? Once upon a time I submitted a bug report that it didn't, and I do not recall the outcome.

garrigue commented:
> Does that actually check the variance? Once upon a time I submitted a bug report that it didn't, and I do not recall the outcome.

Since the call in `transl_with_constraint` has `~check:true` I would expect it to.

gasche committed:
Typemod.merge_constraint: fix computation of the signature environment

(suggested by Leo White and Jacques Garrigue)

gasche commented:
I pushed a new commit that computes and uses `full_env` in the way @lpw25 and @garrigue suggested. (`merge` does not need an `env` parameter anymore.)

gasche commented:
Note: I just confirmed that the regression tests proposed by @garrigue (for constraints / `type_params` unification) and @lpw25 (for `Env.add_item`) are indeed failing with the initial version of the patch.

garrigue commented:
Note that removing the `env` parameter to `merge` is not strictly equivalent: a constraint may result in an improved variance / immediacy / separability, that could be used in other definitions. On the other hand, this would be propagated to other type definitions only in a semi-arbitrary way (i.e. only propagate to following constraints, leaving unconstrained types unchanged, and no fixpoint computation).

garrigue commented:
Actually, the correct code for constrained parameters is a bit different. I will send you a patch in 1 minute.

garrigue commented:
Here is the write code (producing the right error messages).
```ocaml
--- a/typing/typedecl.ml
+++ b/typing/typedecl.ml
@@ -1379,21 +1379,27 @@ let transl_with_constraint env id row_path orig_decl sdecl =
   let params = List.map (fun (cty, _) -> cty.ctyp_type) tparams in
   let orig_decl = Ctype.instance_declaration orig_decl in
   let arity_ok = List.length params = orig_decl.type_arity in
-  if arity_ok then
-    List.iter2 (Ctype.unify_var env) params orig_decl.type_params;
   let constraints = List.map
     (function (ty, ty', loc) ->
-       try
-         let cty = transl_simple_type env false ty in
-         let cty' = transl_simple_type env false ty' in
-         let ty = cty.ctyp_type in
-         let ty' = cty'.ctyp_type in
-         Ctype.unify env ty ty';
-         (cty, cty', loc)
-       with Ctype.Unify tr ->
-         raise(Error(loc, Inconsistent_constraint (env, tr))))
+      let cty = transl_simple_type env false ty in
+      let cty' = transl_simple_type env false ty' in
+      (cty, cty', loc))
     sdecl.ptype_cstrs
   in
+  if arity_ok then
+    List.iter2 
+      (fun (cty, _) -> 
+        try Ctype.unify_var env cty.ctyp_type
+        with Ctype.Unify tr ->
+          raise(Error(cty.ctyp_loc, Inconsistent_constraint (env, tr))))
+      tparams orig_decl.type_params;
+  List.iter
+    (fun (cty, cty', loc) ->
+      try
+        Ctype.unify env cty.ctyp_type cty'.ctyp_type
+      with Ctype.Unify tr ->
+        raise(Error(loc, Inconsistent_constraint (env, tr))))
+    constraints;
   let no_row = not (is_fixed_type sdecl) in
   let (tman, man) =  match sdecl.ptype_manifest with
       None -> None, None
```

garrigue commented:
Here is the correct code (took more than a minute).
```ocaml
--- a/typing/typedecl.ml
+++ b/typing/typedecl.ml
@@ -1379,28 +1379,35 @@ let transl_with_constraint env id row_path orig_decl sdecl =
   let params = List.map (fun (cty, _) -> cty.ctyp_type) tparams in
   let orig_decl = Ctype.instance_declaration orig_decl in
   let arity_ok = List.length params = orig_decl.type_arity in
-  if arity_ok then
-    List.iter2 (Ctype.unify_var env) params orig_decl.type_params;
   let constraints = List.map
     (function (ty, ty', loc) ->
-       try
-         let cty = transl_simple_type env false ty in
-         let cty' = transl_simple_type env false ty' in
-         let ty = cty.ctyp_type in
-         let ty' = cty'.ctyp_type in
-         Ctype.unify env ty ty';
-         (cty, cty', loc)
-       with Ctype.Unify tr ->
-         raise(Error(loc, Inconsistent_constraint (env, tr))))
+      let cty = transl_simple_type outer_env false ty in
+      let cty' = transl_simple_type outer_env false ty' in
+      (cty, cty', loc))
     sdecl.ptype_cstrs
   in
   let no_row = not (is_fixed_type sdecl) in
   let (tman, man) =  match sdecl.ptype_manifest with
       None -> None, None
     | Some sty ->
-        let cty = transl_simple_type env no_row sty in
+        let cty = transl_simple_type outer_env no_row sty in
         Some cty, Some cty.ctyp_type
   in
+  let env = sig_env in
+  if arity_ok then
+    List.iter2 
+      (fun (cty, _) -> 
+        try Ctype.unify_var env cty.ctyp_type
+        with Ctype.Unify tr ->
+          raise(Error(cty.ctyp_loc, Inconsistent_constraint (env, tr))))
+      tparams orig_decl.type_params;
+  List.iter
+    (fun (cty, cty', loc) ->
+      try
+        Ctype.unify env cty.ctyp_type cty'.ctyp_type
+      with Ctype.Unify tr ->
+        raise(Error(loc, Inconsistent_constraint (env, tr))))
+    constraints;
   let priv =
     if sdecl.ptype_private = Private then Private else
     if arity_ok && orig_decl.type_kind <> Type_abstract
```

garrigue commented:
Note that even `transl_simple_type` may end up instantiating some `params`, possibly causing a unification error (pretty improbable, but...)

With this last fix, I think that your code is fine.

And here are examples of inconsistent constraints.
```ocaml
module type Sig = sig type 'a t constraint 'a = int end
module type S = Sig with type 'a t = int constraint 'a = bool
type 'a u constraint 'a = bool
module type S = Sig with type 'a t = int constraint 'a u = 'b
```

gasche commented:
> Note that removing the env parameter to merge is not strictly equivalent: a constraint may result in an improved variance / immediacy / separability, that could be used in other definitions.

It sounds like a good idea to do this in the medium term, although I'm not sure I want to put the effort of learning how to do it *right now*, and also I think it would be nice to have a fix for the issues already identified to let @hhugo go forward in testing more of the JS codebase.

If we wanted to fix this, I think that it would natural to process signature items by recursive groups; we could re-compute type properties (variance, immediacy etc.) for the group modified by the constraint, and then (if a property improved) re-check the following signature items/groups (including submodules, etc.) that may also be improved in trun. But the `signature_item list` we are working on seems to have lost this recursive structure. Is it stored somewhere else, or is there a way to reconstruct it?

lpw25 commented:
> If we wanted to fix this, I think that it would natural to process signature items by recursive groups; we could re-compute type properties (variance, immediacy etc.) for the group modified by the constraint, and then (if a property improved) re-check the following signature items/groups (including submodules, etc.) that may also be improved in trun.

I think a better all around approach would be to just compute these properties on demand (with some laziness/caching). That would also fix the long-standing problems with these properties and functors. Fundamentally, you shouldn't really keep properties on the types if they are not stable under substitution.

As you say though, this is an issue for another day.

garrigue commented:
> If we wanted to fix this, I think that it would natural to process signature items by recursive groups; we could re-compute type properties (variance, immediacy etc.) for the group modified by the constraint, and then (if a property improved) re-check the following signature items/groups (including submodules, etc.) that may also be improved in trun. But the `signature_item list` we are working on seems to have lost this recursive structure. Is it stored somewhere else, or is there a way to reconstruct it?

The `rec_status` field is an hint for recursive blocks, mostly used for printing.
`update_rec_next` maintains it in presence of destructive substitutions.
However, I do agree with @lpw25 that using it for type checking is probably not the way to go, as it only handles very specific cases. And inferred variance information is not useful in signatures, but only on individual types added to the environment, which may be a hint of where it should be recomputed.

gasche commented:
> I think a better all around approach would be to just compute these properties on demand (with some laziness/caching).

I see two natural ways to do this. One is to associate the information to types in the typing environment, computed lazily when the context requires them. The other is to associate the information to types in signatures, computed lazily when the context requires them. The first approach can duplicate computation (if you include a signature several times, you may recompute its type properties several times); but maybe this is fine as we already compute too much today and no one complains. Intuitively I would expect the second to require exactly as much code as the current approach: when we build a new signature from an old one (with small modifications), we have to be careful to recompute the information where it may have changed.

> Fundamentally, you shouldn't really keep properties on the types if they are not stable under substitution.

For me, the "real" object that we are manipulating / talking about is the full derivations (terms with their types, signatures with their properties). It is a standard phenomenon that reducing/computing a derivation may refine its type (the old type is still valid but we can use a more precise one). Whether we want to store the type in the representation of the object itself is a choice of implementation, but I am not convinced that those choices lead to radically different implementations. (If we want to have the types in the representation, they can be forced lazily and be computed by smart constructors to not clutter the main path. If we want to query them after the fact, the query-answering code will look very much like those smart constructors.)

gasche commented:
@garrigue:

> Note that even `transl_simple_type` may end up instantiating some `params`, possibly causing a unification error (pretty improbable, but...)

I'm not sure I understand. Are you suggesting to also wrap each `transl_simple_type` call in a unification-error handler? This is not done by the patch you propose, should the patch be extended to do this?

gasche commented:
@garrigue there is something else that I don't understand with your patch. You are delaying the unification of the `constraints` to happen in the `sig_env`, but this part of the code unifies two types that were themselves translated in `outer_env`, so what was the problem with doing the unification in `outer_env` as well, as the previous code was doing?

gasche commented:
I updated the PR with the fragment that I understand: catching a unification error when unifying `tparams` with `sig_decl.type_params`.

garrigue commented on typing/typedecl.ml:
```diff
@@ -1401,22 +1410,34 @@ let transl_with_constraint env id row_path orig_decl sdecl =
+      try Ctype.unify_var env cty.ctyp_type tparam
+      with Ctype.Unify tr ->
+        raise(Error(cty.ctyp_loc, Inconsistent_constraint (env, tr)))
+    ) tparams sig_decl.type_params;
```
Must also unify the `constraints` you computed above, otherwise what is the point of translating them...

garrigue commented on typing/typedecl.ml:
```diff
@@ -1370,25 +1370,34 @@ let transl_value_decl env loc valdecl =
-         Ctype.unify env ty ty';
          (cty, cty', loc)
        with Ctype.Unify tr ->
          raise(Error(loc, Inconsistent_constraint (env, tr))))
```
No need catch `Unify` if you do not call `unify`.

gasche replied:
Apologies, this was a mistake from partially applying your patch, I of course meant to keep the unification here. I updated the commit with the intended version.

garrigue reviewed (changes requested):
You're dropping constraints...

garrigue commented:
> @garrigue there is something else that I don't understand with your patch. You are delaying the unification of the `constraints` to happen in the `sig_env`, but this part of the code unifies two types that were themselves translated in `outer_env`, so what was the problem with doing the unification in `outer_env` as well, as the previous code was doing?

The point is that unifying the parameters is the main cause to get inconsistent constraints, so this is better to delay the unification until after we have unified the `params`. And if we have unified them, we need to use `sig_env`.

garrigue commented:
> I'm not sure I understand. Are you suggesting to also wrap each `transl_simple_type` call in a unification-error handler? This is not done by the patch you propose, should the patch be extended to do this?

No, this just means if we unify the `params` after calling `transl_simple_type` we may get a unification error, even though doing the unification beforehand would not have caused an error at that point (only when calling `transl_simple_type` later). This doesn't happen much in real code (i.e., having a type with a constrained parameter inside a constraint), but we still need to protect against it.

gasche commented:
> The point is that unifying the parameters is the main cause to get inconsistent constraints, so this is better to delay the unification until after we have unified the params.

If I understand correctly, you are proposing to do the more-likely-to-fail unification first. Could we document the reason for this? (It makes the code more complex.) Is the idea that we get better error messages because they are expressed in terms of the type parameters, rather than the type expressions coming from constraint processing?

Secondary question: we have a few check under a `if arity_ok` test, which in particular avoids exceptions due to length mismatches in `List.{map2,iter2}`. I would suppose that we should also fail if the arities are distinct, but I don't see this anywhere in this function. Is there a check for this somewhere else, or do we support changing the arity in a constraint?

garrigue commented:
> > The point is that unifying the parameters is the main cause to get inconsistent constraints, so this is better to delay the unification until after we have unified the params.
> 
> If I understand correctly, you are proposing to do the more-likely-to-fail unification first. Could we document the reason for this? (It makes the code sensibly more complex.) Is the idea that we get better error messages because they are expressed in terms of the type parameters, rather than the type expressions coming from constraint processing?

Ok, maybe my explanation was not clear. This is the opposite. The goal of unifying the parameters before the constraints is to have unification fail on constraints rather than on parameters. The role of the unification on parameters is to inherit constraints from the original declaration, avoiding having to redeclare constraints by hand. Then, constraint unification will fail if the is a contradiction between the constraints in the original declaration and those added in the `with .. = ... constraint ... = ...` construct. If we do it the other way round, the error will be located on the parameter rather than on the constraint. Both are correct, just this way gives better locations for error messages.
> Secondary question: we have a few check under a `if arity_ok` test, which in particular avoids exceptions due to length mismatches in `List.{map2,iter2}`. I would suppose that we should also fail if the arities are distinct, but I don't see this anywhere in this function. Is there a check for this somewhere else, or do we support changing the arity in a constraint?

After calling `transl_with_constraint`, there is a call to `check_type_decl`, which in turn calls `Includemod.type_declarations`, so the arity will eventually be checked. The idea is to get more uniform error messages. Constraints will also be checked again at that point (in particular, they cannot be more restrictive than in the original declaration). The only reason we may fail on constraints in `transl_with_constraint` is because we want to inherit constraints (and `type_kind`) from the original declaration.

At this point, you may have correctly deduced that writing parameter constraints in with constraints is actually pointless, since anyway you can only inherit the original constraints, and cannot even refine them. I think this is only allowed for backward compatibility reasons (originally there was no constraint inheritance, and there was really no need for `sig_env`). This may also be clearer in some cases.

gasche commented:
Thanks for the explanation. I did as you recommend, delaying the constraint unification to after the parameter unification, with an explanatory comment. That said, I would rather be in favor of doing the constraint unification first, to get simpler code, especially since you explain that using constraints there is in fact useless.

gasche commented:
The diff I propose looks like this:

```diff
     List.map (fun (ty, ty', loc) ->
       let cty = transl_simple_type env false ty in
       let cty' = transl_simple_type env false ty' in
-      (* Note: We delay the unification of those constraints
-         after the unification of parameters, so that clashing
-         constraints report an error on the constraint location
-         rather than the parameter location. *)
+      begin
+        try Ctype.unify_var env cty.ctyp_type tparam
+        with Ctype.Unify tr ->
+          raise(Error(cty.ctyp_loc, Inconsistent_constraint (env, tr)))
+      end;
       (cty, cty', loc)
     ) sdecl.ptype_cstrs
   in
@@ -1424,14 +1425,6 @@ let transl_with_constraint id row_path ~sig_env ~sig_decl ~outer_env sdecl =
       with Ctype.Unify tr ->
         raise(Error(cty.ctyp_loc, Inconsistent_constraint (env, tr)))
     ) tparams sig_decl.type_params;
-  List.iter (fun (cty, cty', loc) ->
-    (* Note: contraints must also be enforced in [sig_env] because
-       they may contain parameter variables from [tparams]
-       that have now be unified in [sig_env]. *)
-    try Ctype.unify env cty.ctyp_type cty'.ctyp_type
-    with Ctype.Unify tr ->
-      raise(Error(loc, Inconsistent_constraint (env, tr)))
-  ) constraints;
```

garrigue reviewed (approved):
Looks good to me, with clear comments.

gasche commented:
Thanks! (I gather that you are *not* in favor of doing the constraint unification earlier to simplify the code?)
I'll try to merge this soon (currently the CI is failing) to go back to 4.11 testing.

gasche commented:
The CI failure is due to a test that was previously failing, and now passes, in `tests/typing-sigsubst/sigsubst.ml`:

```diff
(* Valid substitutions in a recursive module may fail due to the ordering of
   the modules. *)

module type S0 = sig
  module rec M : sig type t = M2.t end
  and M2 : sig type t = int end
end with type M.t = int
 [%%expect {|
-Lines 1-4, characters 17-23:
-1 | .................sig
-2 |   module rec M : sig type t = M2.t end
-3 |   and M2 : sig type t = int end
-4 | end with type M.t = int
-Error: In this `with' constraint, the new definition of M.t
-       does not match its original definition in the constrained signature:
-       Type declarations do not match:
-         type t = int
-       is not included in
-         type t = M2.t
+module type S0 =
+  sig module rec M : sig type t = int end and M2 : sig type t = int end end
 |}]
```

Please correct me if I'm wrong, but my understanding is that this is a welcome change, an improvement due to @lpw25's suggestion to use the whole signature to check the with-constraint.

garrigue commented:
The CI failures seem spurious: I compiled this PR rebased on top of trunk, and `make all-typing` was all green.

gasche commented:
Well I rebased the PR to fix the CI failures, so if you observed the current state of the branch it is natural that it would work. (But I'll still wait for the CI results to merge, and cherry-pick to 4.11.)

garrigue commented:
I believe I pulled and rebased before you pushed the new version (i.e. I had to rebase myself), but I may have missed something.

gasche commented:
Merged, and cherry-picked in 4.11 ( 003dd3993f6b660d2bf99f62b26a6dc2ca558755 ).
